Gt4py is used as a submodule

mcgibbon opened the pull request:
This PR updates pace to use gt4py pinned as a submodule, rather than using GT4PY_VERSION.txt. After this PR, the gt4py caches will use the gt4py sha rather than tag in their GT4PY_VERSION.txt files.

Changes:
- Gt4py is used as a submodule, rather than through GT4PY_VERSION.txt, resulting in the removal of duplicated logic to download gt4py and the addition of `git submodule update --init` commands
- Python used for CircleCI is updated to 3.8, as required by gt4py
- Requirements of gt4py are accounted for when compiling constraints.txt
- constraints.txt is updated
- gt4py is no longer installed as part of daint_venv, and is instead installed from pace
- removed clang-format from requirements, as it is optional and is not available on mac os
- updated tox in fv3core and pace-util to use the gt4py submodule, instead of a hardcoded tag pin

Continued from https://github.com/ai2cm/pace/pull/56.

mcgibbon commented:
> But it will no longer be automatic, this is a little risky

The way we currently update gt4py versions in fv3core/pace is already manual - a commit is made that increments the value in `GT4PY_VERSION.txt`. After this PR, the only thing that fundamentally must change is that instead of manually updating a text file, we manually update a submodule.

If we like, we could still maintain automatic tagging of gt4py commits. We could maintain the current workflow, but have it update pin the submodule to a specific tag if passed in as an environment variable. Or we could make a (very short) workflow that automatically generates and pushes a tag to the gt4py repo of main's submodule commit whenever the main branch is updated.

mcgibbon commented:
launch jenkins

mcgibbon commented:
launch jenkins

mcgibbon commented:
Launched performance test https://jenkins.ginko.ch/job/pace-fv3core-performance-test-PR/21/

mcgibbon commented:
Performance test passed.

mcgibbon commented:
launch jenkins

mcgibbon commented:
It looks like merging this PR will require first launching the cache build plan on it, since the PR involves changing the value used to tag the cache (from gt4py tag to gt4py commit hash).

mcgibbon commented:
Caching plan launched https://jenkins.ginko.ch/view/pace/job/pace-fv3core-cache-setup/141/

mcgibbon commented:
Failed on final copy of caches with an error due to echoing gt4py version to the wrong file location, updated and resubmitting https://jenkins.ginko.ch/view/pace/job/pace-fv3core-cache-setup/143/

mcgibbon commented:
Cancelled cache build job for now, realized the caching infrastructure currently only supports one gt4py tag at a time and that this would disrupt other PRs.

mcgibbon commented:
Updated gt4py caching plans to use multiple tar.gz's for each GT4PY_VERSION. This should avoid breaking other PRs when running the caching plan. Launched caching plan https://jenkins.ginko.ch/view/pace/job/pace-fv3core-cache-setup/144/

mcgibbon commented:
launch jenkins

mcgibbon commented:
Caching plan likely passed since tests passed, but launched it again to see output in case an error happened later down the line. https://jenkins.ginko.ch/view/pace/job/pace-fv3core-cache-setup/158/

mcgibbon commented:
Resubmitting one caching configuration which had a "slurm error", possibly an intermittent issue. https://jenkins.ginko.ch/view/pace/job/pace-fv3core-cache-setup/160/

mcgibbon commented on fv3core/.jenkins/actions/run_standalone.sh:
```diff
@@ -111,13 +117,8 @@ echo "Benchmark directory:          ${BENCHMARK_DIR}"
 echo "Data directory:               ${DATA_DIR}"
 echo "Perf. artifact directory:     ${TIMING_DIR}"
 echo "Profile artifact directory:   ${PROFILE_DIR}"
-echo "Cache directory:              ${CACHE_DIR}"
```
This was removed because CACHE_DIR is not used in this file, and because it gets printed out in the log already from the location where the variable is actually defined and used. Because this file doesn't use the variable, it could be incorrect without causing a failure.

mcgibbon commented on fv3core/.jenkins/actions/fetch_caches.sh:
```diff
@@ -2,24 +2,23 @@
 EXPNAME=$2
 SANITIZED_BACKEND=`echo $BACKEND | sed 's/:/_/g'` #sanitize the backend from any ':'
-CACHE_DIR="/scratch/snx3000/olifu/jenkins/scratch/store_gt_caches/"
+CACHE_DIR="/scratch/snx3000/olifu/jenkins/scratch/store_gt_caches/${EXPNAME}/${SANITIZED_BACKEND}"
```
Here I updated the caching to use a tar.gz that has the gt4py version in its file path, so that we can have simultaneous caches for different gt4py versions. This was needed to be able to test this PR (which involves generating the caches) without breaking other PRs.

jdahm reviewed (approved):
Nice work! There are several parts of this I don't fully understand the logic behind, but those seem to be infrastructure sections that are presumably well-covered by the unit and integration tests. Looks ready to go.

mcgibbon commented:
launch jenkins

mcgibbon commented:
Launched performance test https://jenkins.ginko.ch/job/pace-fv3core-performance-test-PR/35/

mcgibbon commented:
Performance test passed.

twicki commented on external/daint_venv/install.sh:
```diff
@@ -40,23 +36,6 @@ if [ $save_wheel ]; then
-    wget 'https://raw.githubusercontent.com/ai2cm/pace/main/GT4PY_VERSION.txt'
-    GT4PY_VERSION=`cat GT4PY_VERSION.txt`
-fi
-git checkout ${GT4PY_VERSION}
```
removing this checkout of the version means it is not really a relevant variable anymore, correct?
The only use case I see now is for saving /  fetching caches. Does it still make sense to have this as a global variable?

mcgibbon replied:
Yes, I've deleted this everywhere except the fetch/generate caches scripts. This was actually very helpful because it showed me another place we had duplicated the fetch caches logic (which I've now removed).

twicki commented on fv3core/.jenkins/actions/run_standalone.sh:
```diff
@@ -67,7 +67,13 @@ DATA_DIR="/project/s1053/fv3core_serialized_test_data/${DATA_VERSION}/${experime
+SCRIPT_DIR="$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+PACE_DIR=$SCRIPT_DIR/../../../
+
+if [ -z "${GT4PY_VERSION}" ]; then
```
Is having this still relevant?

mcgibbon replied:
No, updated.

twicki commented on fv3core/.jenkins/jenkins.sh:
```diff
@@ -71,10 +71,13 @@ test -f ${BUILDENV_DIR}/machineEnvironment.sh || exitError 1201 ${LINENO} "canno
 
 if [ -z "${GT4PY_VERSION}" ]; then
-    export GT4PY_VERSION=`cat GT4PY_VERSION.txt`
+    export GT4PY_VERSION=`git submodule status ${PACE_DIR}/external/gt4py | awk '{print $1;}'`
```
Is having this still relevant?

mcgibbon replied:
No, updated.

twicki commented on fv3core/examples/standalone/benchmarks/run_on_daint.sh:
```diff
@@ -67,20 +67,20 @@ test -n "$2" || exitError 1002 ${LINENO} "must pass a number of ranks"
+
 if [ -z "${GT4PY_VERSION}" ]; then
-    export GT4PY_VERSION=`cat GT4PY_VERSION.txt`
+    export GT4PY_VERSION=`git submodule status ${PACE_DIR}/external/gt4py | awk '{print $1;}'`
```
Is having this still relevant?

mcgibbon replied:
No, updated.

twicki commented on fv3core/examples/standalone/benchmarks/run_on_daint.sh:
```diff
@@ -122,26 +122,8 @@ echo "    Run arguments:     $run_args"
-        fi
-    fi
-fi
+$FV3CORE_DIR/.jenkins/actions/fetch_caches.sh $backend $EXPNAME
```
:+1:

twicki reviewed (commented):
why is the symlinkf from `pace-util/contstrains.txt` required? In the install we're alyways pointing at the base constraints file, right?

mcgibbon commented on .jenkins/install_virtualenv.sh:
```diff
@@ -11,20 +11,22 @@ exitError()
 
 # check a virtualenv path has been provided
 test -n "$1" || exitError 1001 ${virtualenv_path} "must pass an argument"
+pace_dir=`dirname $0`/../
```
I am not sure that this always works, so I've been using the other pattern wherever I add new operations like this. But I've often kept the old pattern where it exists.

Should I update this to the new pattern?

twicki replied:
I'm ok with leaving it as is. 
I was mostly wondering because we have several new places where you've added pieces along the lines of 
```
SCRIPT_DIR="$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
PACE_DIR=$SCRIPT_DIR/../../../../
```
And we could have used that here as well. But not really necessary

mcgibbon replied:
Updated.

mcgibbon commented on .jenkins/jenkins.sh:
```diff
@@ -73,7 +73,7 @@ echo "PYTHON env ${python_env}"
 
 if [ -z "${GT4PY_VERSION}" ]; then
-    export GT4PY_VERSION=`cat GT4PY_VERSION.txt`
+    export GT4PY_VERSION=`git submodule status ${JENKINS_DIR}/../external/gt4py | awk '{print $1;}'`
```
Yes, this prints the hash.

twicki replied:
:+1:

mcgibbon commented:
> why is the symlinkf from pace-util/contstrains.txt required? In the install we're alyways pointing at the base constraints file, right?

I've fixed it and now removed the symlink. We were not always using the top level file. This is also the case for fv3core, but I'm addressing it in the sphinx docs PR.

twicki commented:
sounds great, thanks

twicki reviewed (approved):
awesome, thanks a lot

mcgibbon commented:
Launched performance test https://jenkins.ginko.ch/job/pace-fv3core-performance-test-PR/37/.
